File: yunva17/Week_19/BOJ_14621_나만안되는연애/BOJ_14621_나만안되는연애.js
let input = require('fs')
  // .readFileSync('/dev/stdin')
  .readFileSync('Week_19/BOJ_14621_나만안되는연애/input.txt')
  .toString()
  .trim()
  .split('\n');

const [N, M] = input[0].split(' ').map(Number);
const gender = input[1].split(' ');

const arr = input
  .slice(2)
  .map((x) => x.split(' ').map(Number))
  .sort((a, b) => a[2] - b[2]);

const parents = Array.from(Array(N + 1), () => -1);

// 유니온 파인드
const find = (a) => {
  if (parents[a] < 0) return a;

  return (parents[a] = find(parents[a]));
};

const union = (a, b) => {
  let rootA = find(a);
  let rootB = find(b);

  if (rootA === rootB) {
    return false;
  }

  if (rootA < rootB) {
    parents[rootA] += parents[rootB];
    parents[rootB] = rootA;
  } else {
    parents[rootB] += parents[rootA];
    parents[rootA] = rootB;
  }

  return true;
};

let answer = 0;
let count = 0;

for (const [u, v, d] of arr) {
  if (gender[u - 1] !== gender[v - 1]) {
    if (union(u, v)) {
      answer += d;

      if (++count === N - 1) break;
    }
  }
}

console.log(count === N - 1 ? answer : -1);

File: hyunseok0314/Week_19/BOJ_14621_나만_안되는_연애.py
import sys

input = sys.stdin.readline

N, M = map(int, input().rstrip().split())

gender = input().rstrip().split()

adj = [[] for _ in range(N)]
edges = []

for _ in range(M):
    u, v, w = map(int, input().rstrip().split())
    u, v = u - 1, v - 1
    if gender[u] != gender[v]:
        adj[u].append(v)
        adj[v].append(u)
        edges.append((w, u, v))

q = [0]
visited = [False for _ in range(N)]
visited[0] = True

count = 1

while q:
    n = q.pop()
    for nn in adj[n]:
        if not visited[nn]:
            q.append(nn)
            visited[nn] = True
            count += 1

if count != N:
    print(-1)
    exit()

parent = [i for i in range(N)]


def find(u):
    if u != parent[u]:
        parent[u] = find(parent[u])
    return parent[u]


def union(u, v):
    u, v = find(u), find(v)
    parent[v] = u


edges.sort()

total = 0

for w, u, v in edges:
    if find(u) != find(v):
        union(u, v)
        total += w

print(total)
